### PR TITLE
[#461] Capacity and increase/decrease policies

### PR DESCRIPTION
**Pull request by maeste**

### CODE REVIEW

File: core/src/main/java/org/ironjacamar/core/connectionmanager/CapacityFiller.java [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the Eclipse
+ *  Public License for more details.
+ *
+ *  You should have received a copy of the Eclipse Public License
+ *  along with this software; if not, write to the Free
+ *  Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ *  02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.ironjacamar.core.connectionmanager;
```
Comment (jesperpedersen) [inline core/src/main/java/org/ironjacamar/core/connectionmanager/CapacityFiller.java]: Better put this under connectionmanager.pool

File: core/src/main/java/org/ironjacamar/core/connectionmanager/CapacityRequest.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the Eclipse
+ *  Public License for more details.
+ *
+ *  You should have received a copy of the Eclipse Public License
+ *  along with this software; if not, write to the Free
+ *  Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ *  02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+
+package org.ironjacamar.core.connectionmanager;
```
Comment (jesperpedersen) [inline core/src/main/java/org/ironjacamar/core/connectionmanager/CapacityRequest.java]: connectionmanager.pool

File: core/src/main/java/org/ironjacamar/core/connectionmanager/SecurityActions.java [hunk 3]
```diff
@@ -148,4 +148,50 @@ public Boolean run()
 
       return equals.booleanValue();
    }
+
```
Comment (jesperpedersen) [inline core/src/main/java/org/ironjacamar/core/connectionmanager/SecurityActions.java]: Move to connectionmanager.pool

File: core/src/main/java/org/ironjacamar/core/connectionmanager/pool/dflt/DefaultManagedConnectionPool.java [hunk 4]
```diff
@@ -286,6 +317,70 @@ else if (cl.getState() == DESTROY)
    }
 
    /**
+    * {@inheritDoc}
+    */
+   @Override
+   public void increaseCapacity()
+   {
+      // We have already created one connection when this method is scheduled
```
Comment (jesperpedersen) [inline core/src/main/java/org/ironjacamar/core/connectionmanager/pool/dflt/DefaultManagedConnectionPool.java]: I think we should also check for prefill credentials here.

Only the prefill credential MCP should be affected by the incrementer

File: core/src/main/java/org/ironjacamar/core/connectionmanager/pool/dflt/DefaultManagedConnectionPool.java [hunk 5]
```diff
@@ -286,6 +317,70 @@ else if (cl.getState() == DESTROY)
...
+               .shouldCreate(currentSize, pool.getConfiguration().getMaxSize(), created);
+
+         if (create)
+         {
+            try
+            {
+               ConnectionListener cl = pool.createConnectionListener(credential, this);
+
+               if (Tracer.isEnabled())
+                  Tracer.createConnectionListener(pool.getType(), this, cl, cl.getManagedConnection(), false, false,
```
Comment (jesperpedersen) [inline core/src/main/java/org/ironjacamar/core/connectionmanager/pool/dflt/DefaultManagedConnectionPool.java]: It is getId() of the pool -- see the other Tracer invocations

File: testsuite/src/test/java/org/ironjacamar/rars/ResourceAdapterFactory.java [hunk 6]
```diff
@@ -557,18 +558,87 @@ else if (tsl == TransactionSupportLevel.LocalTransaction)
...
 
+
+
+   /**
+    * Create the unified-security.rar
+    *
+    *
+    * @return The resource adapter archive
+    */
+   public static ResourceAdapterArchive createCapacityRar()
```
Comment (jesperpedersen) [inline testsuite/src/test/java/org/ironjacamar/rars/ResourceAdapterFactory.java]: I would just add extra parameters to the standard unified security configuration to make it explicit that it is unified.rar that is used

File: core/src/main/java/org/ironjacamar/core/connectionmanager/pool/capacity/CapacityFactory.java [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+         }
+      }
+
+      if (incrementer == null)
+         incrementer = DefaultCapacity.DEFAULT_INCREMENTER;
+
+      // Decrementer
+      if (metadata.getDecrementer() != null && metadata.getDecrementer().getClassName() != null)
+      {
+         if (!isCRI)
```
Comment (jesperpedersen) [inline core/src/main/java/org/ironjacamar/core/connectionmanager/pool/capacity/CapacityFactory.java]: We shouldn't need the isCRI parameter any more, as the checks will be performed in the MCP impl

File: core/src/main/java/org/ironjacamar/core/connectionmanager/pool/dflt/DefaultManagedConnectionPool.java [hunk 8]
```diff
@@ -100,11 +110,25 @@ public DefaultManagedConnectionPool(DefaultPool pool, Credential credential)
     */
    public ConnectionListener getConnectionListener() throws ResourceException
    {
+
       long timestamp = System.currentTimeMillis();
       while (System.currentTimeMillis() - timestamp <= pool.getConfiguration().getBlockingTimeout())
       {
-         for (ConnectionListener cl : listeners)
+         final Iterator<ConnectionListener> listenersIterator;
+         if (pool.isFIFO() && credential.equals(pool.getPrefillCredential()))
```
Comment (jesperpedersen) [inline core/src/main/java/org/ironjacamar/core/connectionmanager/pool/dflt/DefaultManagedConnectionPool.java]: Take this check to the constructor (AbstractMCP), as it is expensive, and set a protected boolean

File: core/src/main/java/org/ironjacamar/core/connectionmanager/pool/dflt/DefaultManagedConnectionPool.java [hunk 9]
```diff
@@ -594,4 +724,13 @@ public ConnectionListener removeConnectionListener(boolean free)
    {
       return removeConnectionListener(free, listeners);
    }
+
```
Comment (jesperpedersen) [inline core/src/main/java/org/ironjacamar/core/connectionmanager/pool/dflt/DefaultManagedConnectionPool.java]: Remove this method, and use Injection to set the field in the necessary test cases.

lastIdleCheck can be moved to AbstractMCP too

File: core/src/main/java/org/ironjacamar/core/connectionmanager/pool/stable/StableManagedConnectionPool.java [hunk 10]
```diff
@@ -62,6 +69,10 @@
     */
    private ConcurrentLinkedDeque<ConnectionListener> listeners;
 
+   /** last idle check */
+   private long lastIdleCheck;
```
Comment (jesperpedersen) [inline core/src/main/java/org/ironjacamar/core/connectionmanager/pool/stable/StableManagedConnectionPool.java]: Same as DefaultMCP

File: core/src/main/java/org/ironjacamar/core/connectionmanager/pool/stable/StableManagedConnectionPool.java [hunk 11]
```diff
@@ -106,8 +117,22 @@ public ConnectionListener getConnectionListener() throws ResourceException
       {
          if (pool.getRequestSemaphore().tryAcquire(pool.getConfiguration().getBlockingTimeout(), TimeUnit.MILLISECONDS))
          {
-            for (ConnectionListener cl : listeners)
+            final Iterator<ConnectionListener> listenersIterator;
+            if (pool.isFIFO() && credential.equals(pool.getPrefillCredential()))
```
Comment (jesperpedersen) [inline core/src/main/java/org/ironjacamar/core/connectionmanager/pool/stable/StableManagedConnectionPool.java]: Same as DefaultMCP

File: testsuite/src/test/java/org/ironjacamar/core/connectionmanager/pool/dflt/CapacityNoCredentialsTestCase.java [hunk 12]
```diff
@@ -0,0 +1,397 @@
...
+      assertNotNull(c1);
+      assertNull(c1.getUserName());
+      assertNull(c1.getPassword());
+
+      assertEquals(1, mcps.size());
+      assertEquals(20, TestUtils.collectionSizeTenSecTimeout(listeners, 20));
+
+      c.close();
+      c1.close();
+      assertEquals(noTxCf.getConnection().getListeners().get(0), listeners.getFirst());
```
Comment (jesperpedersen) [inline testsuite/src/test/java/org/ironjacamar/core/connectionmanager/pool/dflt/CapacityNoCredentialsTestCase.java]: I think it is better to get an identity of the associated ConnectionEventListener through the UnifiedSecurityConnection interface, and check against "listeners". E.g. not expose the ConnectionEventListener directly - same category as ManagedConnection

**Comment by jesperpedersen** [comment]
i18n/l10n logging is available, so CoreLogger / CoreBundle can be used

**Comment by jesperpedersen** [comment]
Rebase

File: core/src/main/java/org/ironjacamar/core/connectionmanager/pool/dflt/DefaultManagedConnectionPool.java (hunk 9 above)
Comment (maeste) [inline core/src/main/java/org/ironjacamar/core/connectionmanager/pool/dflt/DefaultManagedConnectionPool.java]: this method is used also by AbstractPool, not only tests

**Pull request merged**